src/modules/auth/auth.service.ts

```diff
@@ -82,13 +89,77 @@ export class AuthService {
     const accessToken = await this.generateAuthToken(payload, 'access');
     const refreshToken = await this.generateAuthToken(payload, 'refresh');
 
+    const refreshTokenEntity = this.createRefreshTokenEntity(refreshToken);
+    await this.refreshTokenRepository.insert(refreshTokenEntity);
+
     return {
       message: 'User logged in successfully',
       accessToken,
       refreshToken,
     };
   }
 
+  async issueNewTokens(refreshToken: string) {
+    let decodedRefreshToken: AuthTokenPayload;
+
+    try {
+      decodedRefreshToken = await this.jwtService.verifyAsync<AuthTokenPayload>(
+        refreshToken,
+        { secret: this.configService.get<string>('JWT_REFRESH_TOKEN_SECRET') },
+      );
+    } catch (error: unknown) {
+      if (error instanceof TokenExpiredError) {
+        decodedRefreshToken =
+          this.jwtService.decode<AuthTokenPayload>(refreshToken);
+        this.refreshTokenRepository.update(
+          {
+            userId: decodedRefreshToken.sub,
+            token: refreshToken,
+          },
+          { isDeleted: true },
+        );
+      }
+      throw new UnauthorizedException('Invalid or expired token');
+    }
+
+    const existingRefreshTokenEntity =
+      await this.refreshTokenRepository.findOne({
+        where: { userId: decodedRefreshToken.sub, token: refreshToken },
+        relations: {
+          user: true,
+        },
+      });
+
+    if (
+      !existingRefreshTokenEntity ||
+      existingRefreshTokenEntity.isDeleted ||
+      existingRefreshTokenEntity.user.isDeleted
+    ) {
+      throw new UnauthorizedException('Invalid or expired token');
+    }
+
+    const { user } = existingRefreshTokenEntity;
+    const payload = {
+      sub: user.id,
+      userEmail: user.email,
+      userRole: user.role,
+    };
+    const newAccessToken = await this.generateAuthToken(payload, 'access');
+    const newRefreshToken = await this.generateAuthToken(payload, 'refresh');
+
+    const newRefreshTokenEntity =
+      this.createRefreshTokenEntity(newRefreshToken);
+
+    existingRefreshTokenEntity.token = newRefreshToken;
+    existingRefreshTokenEntity.expiresAt = newRefreshTokenEntity.expiresAt;
+    await this.refreshTokenRepository.save(existingRefreshTokenEntity);
```

Comment:
Why you are saving existing one?

Reply:
I was updating the existing one's token, instead of deleting and creating new one. Should I mark existing one as deleted and save a new one?

Reply:
Just hard delete or soft delete the existing one. Don't overthink the solution, make it simple)